Add options for case insensitive properties and enum values.

MikeDombo opened the pull request:
Adds options for case insensitive bean properties and enum values.
This change will close #80.

Added new tests to verify this functionality which are passing. All existing tests are passing.

As this is my first PR on this code base please forgive any issues and let me know what needs to be done to address them. I'm more than happy to work with you to get this in. Our team is trying to migrate to jackson-jr from databind to reduce our memory requirements and we've seen a large impact already.

cowtowncoder commented on jr-annotation-support/src/main/java/com/fasterxml/jackson/jr/annotationsupport/AnnotationBasedIntrospector.java:
```diff
@@ -226,7 +240,7 @@ protected void _findFields() {
     protected void _findFields(final Class<?> currType)
     {
-        if (currType == null || currType == Object.class) {
+        if (currType == null || currType == Object.class || currType == Enum.class) {
```
Oh, actually might be just optimization to get rid of later `isEnumConstant` check?

MikeDombo replied:
And for the purpose of my changing (making `@JsonProperty` work, we of course can't ignore the enum constants), so that explains the change on L240

cowtowncoder replied:
Ok. Gotcha.

cowtowncoder commented on jr-annotation-support/src/main/java/com/fasterxml/jackson/jr/annotationsupport/AnnotationBasedValueRWModifier.java:
```diff
@@ -32,4 +38,40 @@ public POJODefinition pojoDefinitionForSerialization(JSONWriter writeContext,
+
+        @Override
+        public void writeValue(JSONWriter context, JsonGenerator g, Object value) throws IOException {
+            // Try getting the value from the JsonProperty annotation, if it fails just use the enum name.
```
This introspection should be done once when constructing writer, and not each every time value is written -- introspection access is pretty slow.

cowtowncoder commented on jr-objects/src/main/java/com/fasterxml/jackson/jr/ob/impl/ValueWriterLocator.java:
```diff
@@ -307,6 +309,16 @@ private int _registerWriter(Class<?> rawType, ValueWriter valueWriter) {
                 }
             }
             int typeId = _findSimpleType(type, true);
+            // Give plugin the opportunity to override standard value writer
```
Ah. So this call was missing?

MikeDombo replied:
Yes, the override was never getting called at all for enums specifically. I suspect it is because Bean properties are treated differently than root-level types.

cowtowncoder commented on jr-objects/src/main/java/com/fasterxml/jackson/jr/ob/impl/ValueReaderLocator.java:
```diff
@@ -295,10 +304,27 @@ protected ValueReader arrayReader(Class<?> contextType, Class<?> arrayType) {
+                try {
+                    byName.put(e.name, e.field.get(null));
+                } catch (IllegalAccessException ignored) {
+                }
```
Is this safe to ignore? Or would it cause hard-to-debug issues? (if access fails, code really can't succeed)

MikeDombo replied:
I don't believe that enum values are able to be privated, so I'm fairly sure that this exception is impossible. If I did want to raise it up, how would that be done from this part of the code base? Just using an unchecked exception?

MikeDombo commented:
Thank you for the initial round of feedback. I'll work on it again tomorrow to split it up, rebase on 2.13, and undo my IDE formatting.

MikeDombo commented:
Updated and split into #82 and #83.
CLA should be covered by existing CCLA from Amazon/Amazon Web Services, I'm following up with legal to make sure everything is good with that.

cowtowncoder commented:
Very cool, thank you for covering everything I brough up. Yes, C(C)LA is fine.
I hope to go over the changes very soon now: this is on top of my todo list. This week has been bit hectic (and I will be traveling next week), but I should be able to get these merged during this week.

I really appreciate help with jackson-jr: I haven't had quite as much time to look into improving it due to other Jackson related things. But I think there is a lot of potential, and due to smaller size it is still quite a bit easier to make targeted improvements compared to "full" Jackson.
So I want to avoid being a big bottleneck wrt PRs; but at the same time obviously want to understand changes instead of just merging everything (both to help sanity check some global aspects and keep my understanding of the whole up to date).

MikeDombo commented:
Rebasing and ensuring every test is passing.

MikeDombo commented:
Rebased and all tests are passing

cowtowncoder commented on jr-objects/src/main/java/com/fasterxml/jackson/jr/ob/impl/ValueReaderLocator.java:
```diff
@@ -441,7 +469,8 @@ protected BeanReader _resolveBeanForDeser(Class<?> raw, POJODefinition beanDef)
             propMap = Collections.emptyMap();
         } else {
-            propMap = new HashMap<String, BeanPropertyReader>();
+            propMap = JSON.Feature.ACCEPT_CASE_INSENSITIVE_PROPERTIES.isEnabled(_features) ? new TreeMap<String,
```
(and yes, after googling and reading about it found out that's exactly what happens)

cowtowncoder reviewed (commented):
Looks good, will merge!
